Sources/Testing/Test.Case.Generator.swift

```diff
@@ -93,34 +117,134 @@ extension Test.Case {
+    ///     cases should be generated.
+    ///   - parameters: The parameters of the test function for which test cases
+    ///     should be generated.
+    ///   - testFunction: The test function to which each generated test case
+    ///     passes an argument value from `collection`.
+    ///
+    /// This initializer overload is specialized for collections of 2-tuples to
+    /// efficiently de-structure their elements when appropriate.
+    ///
+    /// @Comment {
+    ///   - Bug: The testing library should support variadic generics.
+    ///     ([103416861](rdar://103416861))
+    /// }
+    init<E1, E2>(
+      arguments collection: S,
+      parameters: [Test.ParameterInfo],
+      testFunction: @escaping @Sendable ((E1, E2)) async throws -> Void
+    ) where S: Collection, S.Element == (E1, E2) {
+      self.init(sequence: collection, parameters: parameters, testFunction: testFunction)
+    }
+
     /// Initialize an instance of this type that iterates over the specified
     /// zipped sequence of argument values.
     ///
     /// - Parameters:
     ///   - zippedCollections: A zipped sequence of argument values for which
     ///     test cases should be generated.
+    ///   - parameters: The parameters of the test function for which test cases
+    ///     should be generated.
     ///   - testFunction: The test function to which each generated test case
     ///     passes an argument value from `zippedCollections`.
     init<C1, C2>(
       arguments zippedCollections: Zip2Sequence<C1, C2>,
+      parameters: [Test.ParameterInfo],
       testFunction: @escaping @Sendable ((C1.Element, C2.Element)) async throws -> Void
-    ) where S == Zip2Sequence<C1, C2> {
-      self.init(sequence: zippedCollections) { element in
-        Test.Case(arguments: [element]) {
-          try await testFunction(element)
+    ) where S == Zip2Sequence<C1, C2>, C1: Collection, C2: Collection {
+      self.init(sequence: zippedCollections, parameters: parameters, testFunction: testFunction)
+    }
+
+    /// Initialize an instance of this type that iterates over the specified
+    /// dictionary of argument values.
+    ///
+    /// - Parameters:
+    ///   - dictionary: A dictionary of argument values for which test cases
+    ///     should be generated.
+    ///   - parameters: The parameters of the test function for which test cases
+    ///     should be generated.
+    ///   - testFunction: The test function to which each generated test case
+    ///     passes an argument value from `dictionary`.
+    ///
+    /// This initializer overload is specialized for dictionary collections, to
+    /// efficiently de-structure their elements (which are known to be 2-tuples)
+    /// when appropriate.
+    init<Key, Value>(
+      arguments dictionary: Dictionary<Key, Value>,
+      parameters: [Test.ParameterInfo],
+      testFunction: @escaping @Sendable ((Key, Value)) async throws -> Void
+    ) where S == Dictionary<Key, Value> {
```

Comment:
Dictionaries are sequences of 2-tuples already, so is this overload needed in _addition_ to the 2-tuple overload? Seems like the decomposition of a dictionary here is exactly the same as the decomposition of a tuple (either into a single tuple argument or into _n_ desugared arguments.)

Reply:
I thought it was redundant also but the compiler treats `Dictionary` differently because its `Element` type is `(key: Key, value: Value)` and not `(Key, Value)`. The tuple labels make it not identical, in other words.

Reply:
Boooo-urns. Okay. Can you add a comment about that then please?

Reply:
(And add it in Test+Macro.swift too? Actually… I wonder if we've specialized `__function()` enough at this point that we could (in a different PR!) refactor this type and move the logic into that file, reducing redundancy?

Reply:
Sure, will note that in docs